src/rust/engine/stdio/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+        // Write to the underlying Console.
+        let res = console.write_stderr(content);
+        if res.is_ok() {
+          return;
+        }
+        // If writing to the stdout handle fails, fall through to mutate self to drop it.
+        res.map_err(|e| e.to_string())
+      }
+      InnerDestination::Exclusive { ref stderr_handler } => {
+        // Write to the Exclusive handler.
+        let res = stderr_handler(&String::from_utf8_lossy(content));
+        if res.is_ok() {
+          return;
+        }
+        // If writing to the stderr handler fails, fall through to clear it and try again.
+        res.map_err(|()| "Failed to write stderr to Exclusive".to_owned())
+      }
+      InnerDestination::Logging => {
+        // Release the lock on the Destination before logging.
+        std::mem::drop(destination);
+        log::info!("stderr: {:?}", String::from_utf8_lossy(content));
+        return;
+      }
+    };
+
+    // Release the lock, clear the Console, log the error and retry.
+    let error_str = format!(
+      "Failed to write stderr to {:?}, falling back to Logging: {:?}",
+      destination, error_res
+    );
+    std::mem::drop(destination);
+    self.console_clear();
+    log::warn!("{}", error_str);
+    self.write_stderr(content);
+  }
+
+  ///
+  /// If stdin is backed by a real file, returns it as a RawFd. All usage of `RawFd` is unsafe,
+  /// but this method is additionally unsafe because the real file might have been closed by the
+  /// time the caller interacts with it.
+  ///
+  pub fn stdin_as_raw_fd(&self) -> Result<RawFd, String> {
+    match &*self.0.lock() {
+      InnerDestination::Console(console) => Ok(console.stdin_as_raw_fd()),
+      InnerDestination::Logging => {
+        Err("No associated file descriptor for the Logging destination".to_owned())
+      }
+      InnerDestination::Exclusive { .. } => {
+        Err("A UI or process has exclusive access, and must be stopped before stdio is directly accessible.".to_owned())
+      }
+    }
+  }
+
+  ///
+  /// If stdout is backed by a real file, returns it as a RawFd. All usage of `RawFd` is unsafe,
+  /// but this method is additionally unsafe because the real file might have been closed by the
+  /// time the caller interacts with it.
+  ///
+  pub fn stdout_as_raw_fd(&self) -> Result<RawFd, String> {
+    match &*self.0.lock() {
+      InnerDestination::Console(console) => Ok(console.stdout_as_raw_fd()),
+      InnerDestination::Logging => {
+        Err("No associated file descriptor for the Logging destination".to_owned())
+      }
+      InnerDestination::Exclusive { .. } => {
+        Err("A UI or process has exclusive access, and must be stopped before stdio is directly accessible.".to_owned())
+      }
+    }
+  }
+
+  ///
+  /// If stdout is backed by a real file, returns it as a RawFd. All usage of `RawFd` is unsafe,
+  /// but this method is additionally unsafe because the real file might have been closed by the
```

Comment:
This comment is kind of confusing, since this method is not marked as a Rust `unsafe` function. IIRC the `libc`  functions that manipulate `RawFd` are generally unsafe, but that's not the same thing as this function that generates `RawFd` being unsafe.

Reply:
The unsafety here is technically just the same unsafety that is possible for any `RawFd`... this function will not directly panic: instead, the `RawFd` that you get might already be closed. So the comment is probably misleading, yea.